### PR TITLE
11 integrate the GitHub action ci with cat-github-actions

### LINKED ISSUE #69: CI is currently not working...  Looks bad
Go to the actions TAB, everything is failing...
Actions -> All Workflows.
image

All pushes PR are not unvetted.. and accepted in good will of reviewers... There is a no checks at all... cos the original tests are too old to run...
image

### LINKED ISSUE #72: Refactor, extend and develop to integrate the GitHub Action CI with https://github.com/puppetlabs/cat-github-actions/tree/main/.github
## Steps to Reproduce
1. Check in codes or anything.
2. Trigger the CI github action
3. Go to `Actions` above, and the tests fail.

## Environment
 - Version : PE 2021.7.8
 - Platform :  Implemented only the ourdated `ubuntu-16.04`
             Supposed to Oracle Linux 7, RedHat Linux 7 and 8, Ubuntu 18.04, 20.04 and 22.04
 
## Additional Context
Ref:  https://github.com/sooyean-hoo/puppetlabs-servicenow_cmdb_integration/issues/11

Adjust the CI to make it 100% with https://github.com/puppetlabs/cat-github-actions/tree/main/.github . 

 Firstly, I am **not the original author** of the Acceptance Tests. It took me 
 weeks to figure out what is the design and the original author's intentions. 
 Have fun using 'git blame'.

 **The original Acceptance** Test is using VMPooler... so it run the VMPooler to 
 get a machine via SSH... as master. Inside the master, you run a mockup 
 ServiceNow in a docker to test against. All acceptance tests are to be run on 
 the master. VMPooler is now deprecated. Hence I am using vagrant to replace 
 the VMPooler VM, using docker as a replacement has proven to be not working, 
 as it is hard to have a docker within a docker container. In fact, it is 
 impossible.

**Now for the inner working of the tests.** It is also silly to try to change the 
  test to work with 2 containers: too much changes. The original author use 
  litmus to only provision 1 VM: the master.  The rest of the provision, the 
 servicenow, it did it not via Litmus, but it  still want to use it to manage and 
 access the docker. This is where it is confusing. There is a 'servicenow_instance' in 
 spec/support/acceptance/helpers.rb, magically created. It is not mentioned at 
 all in provision.yaml. It is not created by the litmus, but a script. I would 
 called it, as a way to hack the litmus. Hence it is as good as how litmus stay
 constant. If there is any future changes to how litmus ingest the inventory. 
 It may just fail. Good LUCK.

---
Also not sure why, when you provision a VM via Provision the module (for Vagrant, VirtualBox) or indirectly via Litmus, it provision you a VM which you cannot ssh into. I do not have time to study that. But it is just too stupid. In the end I had to write a bunch script to compensate.

---
This module's original  acceptance is poorly written. I would argue that it is totally not written. 

The only acceptance ever written is for Ubuntu 16.04... using VMPool. Nothing else is considered..... Even though it is stated to work on a lot more.
ref: https://github.com/sooyean-hoo/puppetlabs-servicenow_cmdb_integration/blob/a4c76f6b718ac9b9ee824324a904ece15b04432d/metadata.json#L24

```json
  "operatingsystem_support": [
    {
      "operatingsystem": "CentOS",
      "operatingsystemrelease": [
        "7",
        "8"
      ]
    },
    {
      "operatingsystem": "OracleLinux",
      "operatingsystemrelease": [
        "7",
        "8"
      ]
    },
    {
      "operatingsystem": "RedHat",
      "operatingsystemrelease": [
        "7",
        "8"
      ]
    },
    {
      "operatingsystem": "Scientific",
      "operatingsystemrelease": [
        "7",
        "8"
      ]
    },
    {
      "operatingsystem": "Ubuntu",
      "operatingsystemrelease": [
        "18.04"
      ]
    },
    {
      "operatingsystem": "SLES",
      "operatingsystemrelease": [
        "12"
      ]
    }
  ]
```

The line which draw the conclusion that the acceptance is only for Ubuntu is @ here.  https://github.com/sooyean-hoo/puppetlabs-servicenow_cmdb_integration/blob/a4c76f6b718ac9b9ee824324a904ece15b04432d/spec/support/acceptance/install_pe.sh#L8

The file install_pe.sh is used in all variant, hence all tests will failed for now, esp. even O/S Ubuntu 16.04 is not longer supported.

The task originally was to just to update the tests to bring it to date. However now it turned out to be a bigger task: writing test to work with all the missing O/S. about 10 of them. This is way more complicated, as there is also need to add more scenario and cater/reimplement for all old functions and original testing goals in the new O/S.
  
  ---
  https://github.com/sooyean-hoo/puppetlabs-servicenow_cmdb_integration/actions/runs/10669025799

image

So now only *Oracle Linux 7*, *RedHat Linux 7 and 8*, *Ubuntu 18.04, 20.04 and 22.04* are
 part of the testing Matrix. The exclusions and the reason for it are listed below.
 As an overview, there are a lot difficult in running the mock-up `ServiceNow server`.
 It is still run as a docker, but for Ubuntu, they are hosted in the `PE Primary server`.
 For others, they are hosted in the Runner. This is because I only managed to get docker to
 play nice in Ubuntu. And yes the Runner is also of Ubuntu.
 There are playful manipulation of port forwardings and /etc/hosts used to standardise 
 the access to both the `PE primary server `and the` mock-up ServiceNow server`.

### PR DESCRIPTION
**Pull request by sooyean-hoo**
- Ref: 
> https://github.com/sooyean-hoo/puppetlabs-servicenow_cmdb_integration/issues/11 
> https://github.com/puppetlabs/puppetlabs-servicenow_cmdb_integration/issues/72 #72 

TLDR; 

The line which draw the conclusion that the acceptance is only for Ubuntu is @ here.  https://github.com/sooyean-hoo/puppetlabs-servicenow_cmdb_integration/blob/a4c76f6b718ac9b9ee824324a904ece15b04432d/spec/support/acceptance/install_pe.sh#L8

The file install_pe.sh is used in all variant, hence all tests will failed for now, esp. even O/S Ubuntu 16.04 is not longer supported.

The task originally was to just to update the tests to bring it to date. However now it turned out to be a bigger task: writing test to work with all the missing O/S. about 10 of them. This is way more complicated, as there is also need to add more scenario and cater/reimplement for all old functions and original testing goals in the new O/S.

Now. 
So now only *Oracle Linux 7*, *RedHat Linux 7 and 8*, *Ubuntu 18.04, 20.04 and 22.04* are
 part of the testing Matrix. The exclusions and the reason for it are listed below.
 As an overview, there are a lot difficult in running the mock-up `ServiceNow server`.
 It is still run as a docker, but for Ubuntu, they are hosted in the `PE Primary server`.
 For others, they are hosted in the Runner. This is because I only managed to get docker to
 play nice in Ubuntu. And yes the Runner is also of Ubuntu.
 There are playful manipulation of port forwardings and /etc/hosts used to standardise 
 the access to both the `PE primary server `and the` mock-up ServiceNow server`.
 https://github.com/sooyean-hoo/puppetlabs-servicenow_cmdb_integration/actions/runs/10669025799

image

### CODE REVIEW

**Comment by gregohardy** [comment]
Hi @sooyean-hoo. Great stuff. Dont forget to squash your commit so we can review it.

**Comment by sooyean-hoo** [comment]
@gregohardy I squashed it before... to  a few commits actually... Anyway with the recent update, I squashed it further to only 1 commit... 

Along with the squash, I add a new O/S to the matrix. I cannot explain why I can get OracleLinux 7 and RHEL-8 working, but not OracleLinux 8....... 

I was also puzzled why I wrote... https://github.com/sooyean-hoo/puppetlabs-servicenow_cmdb_integration/blob/386d71745ace9395d91a7d09ac5bee95ec393a97/.github/workflows/ci_proposed.yml#L78

```
# \"OracleLinux-8\", = ?????
```

Hence I did a review on OracleLinux 8. It turns out that `matrix_from_metadata_v2`  @ https://github.com/puppetlabs/puppet_litmus/blob/v0.36.1/exe/matrix_from_metadata_v2 (version 0.36.1) will reject OracleLinux 8 as it cannot find a match.

Hence I cannot translate it into a vagrant VM. To which I need to do some 'magic' to include it.

# **Long Story short, I squashed the commits into 1 and extended the number of O/Ses in the testing matrix from 6 to 7, with the new O/S as `Oracle Linux 8` and the original 6 O/Ses as  `OracleLinux 7`, `Ubuntu 18.04, 20.04 22.04` and `RedHat 7 & 8` . **

As the time of this comment, the tests are still running. Hence I am not sure how good or bad the results are. Sometimes it needs a bit of nudge aka `re-run the failed one` to work. I did test before in another branch. It was successful for the `OracleLinux 8`. See: https://github.com/sooyean-hoo/puppetlabs-servicenow_cmdb_integration/actions/runs/10679611998/job/29599273899
